src/Nethermind/Nethermind.Optimism/CL/P2P/OptimismCLP2P.cs

```diff
@@ -132,6 +175,64 @@ private async Task MainLoop(CancellationToken token)
         }
     }
 
+    private async Task<ExecutionPayloadV3?> TryRequestPayload
+        (ISession? session, ulong payloadNumber, Hash256 expectedHash, CancellationToken token)
+    {
+        try
+        {
+            CancellationTokenSource cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token);
+            cancellationTokenSource.CancelAfter(TimeSpan.FromSeconds(5));
+            if (session is null) return null;
+            ExecutionPayloadV3? payload =
+                await session.DialAsync<PayloadByNumberProtocol, ulong, ExecutionPayloadV3?>(payloadNumber, cancellationTokenSource.Token);
+            if (payload is null || payload.BlockHash != expectedHash) return null;
+            return payload;
+        }
+        catch (Exception e)
+        {
+            if (_logger.IsTrace) _logger.Trace($"Unable to request payload {payloadNumber} ({expectedHash}). Exception: {e}");
+            return null;
+        }
+    }
+
+    private async Task<ExecutionPayloadV3?> RequestPayload(ulong payloadNumber, Hash256 expectedHash, CancellationToken token)
+    {
+        if (_logger.IsInfo) _logger.Info($"Requesting missing payload. Number: {payloadNumber}, Expected hash: {expectedHash}");
+        try
+        {
+            ExecutionPayloadV3? response = null;
+            foreach (Multiaddress peer in _peerManager.GetPeers())
+            {
+                ISession? remotePeer = await TryDialPeer(peer, token);
+                response = await TryRequestPayload(remotePeer, payloadNumber, expectedHash, token);
+                if (response is not null)
+                {
+                    _peerManager.IncreaseRating(peer);
+                    await remotePeer!.DisconnectAsync();
+                    break;
+                }
+
+                if (remotePeer is not null) await remotePeer.DisconnectAsync();
```

Comment:
Discussion:

There is a pattern of `DialAsync` with a timeout, making a request, and then disconnecting. We might want to create a wrapper for this to avoid possible dangling connections using `IAsyncDisposable`.